src/transformStateWithClones.js

```diff
@@ -7,7 +7,36 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  let currentState = { ...state }; // a shallow clone of the initial state
+  const stateHistory = [];
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        // Add properties from extraData to the current state
+        currentState = { ...currentState, ...action.extraData };
+        break;
+
+      case 'removeProperties':
+        // Remove specified keys from the current state
+        currentState = { ...currentState }; // Create a clone before modificat
+
+        for (const key of action.keysToRemove) {
+          delete currentState[key]; // Remove each key
+        }
+        break;
+
+      case 'clear':
+        // Reset current state to an empty object
+        currentState = {};
+        break;
+    }
+
+    // Push a clone of the updated current state to the history array
+    stateHistory.push({ ...currentState });
```

Comment:
The cloning and pushing of the `currentState` to `stateHistory` should be done outside of the `switch` block, at the end of each loop cycle. This ensures that the state is captured after all actions are processed, as per the additional prompt rules.